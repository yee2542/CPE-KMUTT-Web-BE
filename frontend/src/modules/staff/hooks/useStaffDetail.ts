import { useRouter } from 'next/router';

import { CrumbProps } from 'common/components/Layouts/CommonLayout/types';
import { GetStaffQuery } from 'common/generated/generated-types';

import { BREADCRUMB } from 'modules/staff/pages/StaffsPage/constants';
import { StaffDetail } from 'modules/staff/types';
import { staffDetailMapping } from 'modules/staff/utils/staffDetailMapping';

export type UseStaffDetail = {
  mappedStaff: StaffDetail;
  pageBreadcrumb: CrumbProps[];
};

const useStaffDetail = (data: GetStaffQuery): UseStaffDetail => {
  const router = useRouter();

  const mappedStaff = staffDetailMapping(data);
  const pageBreadcrumb = [
    ...BREADCRUMB,
    {
      title: `${mappedStaff?.name || ''}`,
      link: router.asPath,
    },
  ];

  return {
    mappedStaff,
    pageBreadcrumb,
  };
};

export default useStaffDetail;
