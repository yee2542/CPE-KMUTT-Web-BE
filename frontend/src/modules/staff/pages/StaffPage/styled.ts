import styled from 'styled-components';

import Container from 'common/components/Container';

export const StyledContainer = styled(Container)`
  gap: 30px;
  padding-top: 40px;
  padding-bottom: 40px;

  > :first-child {
    margin-bottom: 10px;
  }
`;
