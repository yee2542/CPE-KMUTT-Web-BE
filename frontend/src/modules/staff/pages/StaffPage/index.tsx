import { FC } from 'react';

import Header from 'common/components/Header';
import CommonLayout from 'common/components/Layouts/CommonLayout';
import { GetStaffQuery } from 'common/generated/generated-types';
import baseUrl from 'common/utils/baseUrl';

import BackSection from 'modules/staff/components/BackSection';
import StaffCommonSection from 'modules/staff/components/StaffCommonContactSection';
import StaffContactSection from 'modules/staff/components/StaffContactSection';
import { StyledTable } from 'modules/staff/components/StaffContactSection/styled';
import useStaffDetail from 'modules/staff/hooks/useStaffDetail';
import { StaffDetail } from 'modules/staff/types';

import { StyledContainer } from './styled';

const HEADER_BG_IMAGE = baseUrl('/images/staff_bg_header.jpg');

const StaffPage: FC<{ data: GetStaffQuery }> = ({ data }) => {
  const { mappedStaff, pageBreadcrumb: PageBreadcrumb } = useStaffDetail(data);

  return (
    <CommonLayout
      navigate={PageBreadcrumb}
      Header={
        <Header
          title="บุคลากร"
          subtitle="Our staffs"
          variant="2-col"
          backgroundImage={HEADER_BG_IMAGE}
        />
      }
      withFaqBanner>
      <StyledContainer>
        <BackSection />
        <StaffContactSection staff={mappedStaff} />
        <StaffInfo mappedStaff={mappedStaff} />
        <StaffSubjects mappedStaff={mappedStaff} />
        <StaffResearch mappedStaff={mappedStaff} />
      </StyledContainer>
    </CommonLayout>
  );
};

type CommonSectionProps = { mappedStaff: StaffDetail };

const StaffResearch: FC<CommonSectionProps> = ({ mappedStaff }) => (
  <StaffCommonSection header="งานวิจัย/หัวข้อที่สนใจ">
    {mappedStaff.info}
  </StaffCommonSection>
);

const StaffInfo: FC<CommonSectionProps> = ({ mappedStaff }) => (
  <StaffCommonSection header="ประวัติ">{mappedStaff.info}</StaffCommonSection>
);

const StaffSubjects: FC<CommonSectionProps> = ({ mappedStaff }) => (
  <StaffCommonSection header="วิชาที่สอน">
    <StyledTable>
      <tr>
        <td>รหัสวิชา</td>
        <td>ชื่อวิชา</td>
      </tr>
      {mappedStaff.subjects.map((s) => (
        <>
          <td>{s.subjectId}</td>
          <td>{s.title}</td>
        </>
      ))}
    </StyledTable>
  </StaffCommonSection>
);

export default StaffPage;
