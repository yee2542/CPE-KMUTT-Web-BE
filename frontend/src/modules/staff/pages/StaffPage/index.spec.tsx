import { shallow } from 'enzyme';
import StaffPage from 'pages/staff/[id]';

import { MOCK_API } from 'modules/staff/hooks/useStaffDetail.spec';

describe('StaffPage /[id] Tests', () => {
  it('Navbar menu should be defined', () => {
    const wrapper = shallow(<StaffPage data={MOCK_API} />);
    expect(wrapper).toMatchSnapshot();
  });
});
