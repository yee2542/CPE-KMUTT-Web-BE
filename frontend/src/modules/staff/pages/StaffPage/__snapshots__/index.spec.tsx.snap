// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffPage /[id] Tests Navbar menu should be defined 1`] = `
<CommonLayout
  Header={
    <Header
      backgroundImage="/images/staff_bg_header.jpg"
      subtitle="Our staffs"
      title="บุคลากร"
      variant="2-col"
    />
  }
  navigate={
    Array [
      Object {
        "link": "/",
        "title": "หนัาหลัก",
      },
      Object {
        "link": "/about-us",
        "title": "เกี่ยวกับเรา",
      },
      Object {
        "link": "/staffs",
        "title": "บุคคลากร",
      },
      Object {
        "link": "",
        "title": "ชื่อจริง นามสกุล",
      },
    ]
  }
  withFaqBanner={true}
>
  <styled__StyledContainer>
    <BackSection />
    <StaffContactSection
      staff={
        Object {
          "academicPosition": "อาจารย์",
          "academicPositionGroup": "",
          "email": "mail@mail.com",
          "fullTitle": "ผู้ช่วยศาสตราจารย์",
          "id": "",
          "imageUrl": "http://localhost:1337/files/test.png",
          "info": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>info</li>
<li>info</li>
<li>info</li>
</ul>",
              }
            }
          />,
          "name": "ชื่อจริง นามสกุล",
          "phone": "0910000000",
          "research": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>research</li>
<li>research</li>
<li>research</li>
<li>&nbsp;</li>
</ul>",
              }
            }
          />,
          "rooms": Array [
            Object {
              "roomId": "CPE1113",
              "title": "ห้องพัก",
            },
          ],
          "subjects": Array [
            Object {
              "subjectId": "CPE101",
              "title": "Computer Programming",
            },
          ],
          "title": "ผศ",
        }
      }
    />
    <StaffInfo
      mappedStaff={
        Object {
          "academicPosition": "อาจารย์",
          "academicPositionGroup": "",
          "email": "mail@mail.com",
          "fullTitle": "ผู้ช่วยศาสตราจารย์",
          "id": "",
          "imageUrl": "http://localhost:1337/files/test.png",
          "info": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>info</li>
<li>info</li>
<li>info</li>
</ul>",
              }
            }
          />,
          "name": "ชื่อจริง นามสกุล",
          "phone": "0910000000",
          "research": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>research</li>
<li>research</li>
<li>research</li>
<li>&nbsp;</li>
</ul>",
              }
            }
          />,
          "rooms": Array [
            Object {
              "roomId": "CPE1113",
              "title": "ห้องพัก",
            },
          ],
          "subjects": Array [
            Object {
              "subjectId": "CPE101",
              "title": "Computer Programming",
            },
          ],
          "title": "ผศ",
        }
      }
    />
    <StaffSubjects
      mappedStaff={
        Object {
          "academicPosition": "อาจารย์",
          "academicPositionGroup": "",
          "email": "mail@mail.com",
          "fullTitle": "ผู้ช่วยศาสตราจารย์",
          "id": "",
          "imageUrl": "http://localhost:1337/files/test.png",
          "info": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>info</li>
<li>info</li>
<li>info</li>
</ul>",
              }
            }
          />,
          "name": "ชื่อจริง นามสกุล",
          "phone": "0910000000",
          "research": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>research</li>
<li>research</li>
<li>research</li>
<li>&nbsp;</li>
</ul>",
              }
            }
          />,
          "rooms": Array [
            Object {
              "roomId": "CPE1113",
              "title": "ห้องพัก",
            },
          ],
          "subjects": Array [
            Object {
              "subjectId": "CPE101",
              "title": "Computer Programming",
            },
          ],
          "title": "ผศ",
        }
      }
    />
    <StaffResearch
      mappedStaff={
        Object {
          "academicPosition": "อาจารย์",
          "academicPositionGroup": "",
          "email": "mail@mail.com",
          "fullTitle": "ผู้ช่วยศาสตราจารย์",
          "id": "",
          "imageUrl": "http://localhost:1337/files/test.png",
          "info": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>info</li>
<li>info</li>
<li>info</li>
</ul>",
              }
            }
          />,
          "name": "ชื่อจริง นามสกุล",
          "phone": "0910000000",
          "research": <div
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ul>
<li>research</li>
<li>research</li>
<li>research</li>
<li>&nbsp;</li>
</ul>",
              }
            }
          />,
          "rooms": Array [
            Object {
              "roomId": "CPE1113",
              "title": "ห้องพัก",
            },
          ],
          "subjects": Array [
            Object {
              "subjectId": "CPE101",
              "title": "Computer Programming",
            },
          ],
          "title": "ผศ",
        }
      }
    />
  </styled__StyledContainer>
</CommonLayout>
`;
